### PR TITLE
Make regexp pattern `[^a]` consistent with Spark for multiline strings

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

Closes https://github.com/NVIDIA/spark-rapids/issues/4229

The following documentation from this PR explains the transpiler change that makes us consistent with CPU for patterns such as `[^a]`.

```
// There are differences between cuDF and Java handling of newlines
// for negative character matches. The expression `[^a]` will match
// `\r` and `\n` in Java but not in cuDF, so we replace `[^a]` with
// `(?:[\r\n]|[^a])`. We also have to take into account whether any
// newline characters are included in the character range.
//
// Examples:
//
// `[^a]`     => `(?:[\r\n]|[^a])`
// `[^a\r]`   => `(?:[\n]|[^a])`
// `[^a\n]`   => `(?:[\r]|[^a])`
// `[^a\r\n]` => `[^a]`
```

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 1]
```diff
@@ -494,7 +494,40 @@ class CudfRegexTranspiler(replace: Boolean) {
         }
         val components: Seq[RegexCharacterClassComponent] = characters
           .map(x => rewrite(x).asInstanceOf[RegexCharacterClassComponent])
-        RegexCharacterClass(negated, ListBuffer(components: _*))
+
+        if (negated) {
+          // There are differences between cuDF and Java handling of newlines
+          // for negative character matches. The expression `[^a]` will match
+          // `\r` and `\n` in Java but not in cuDF, so we replace `[^a]` with
+          // `(?:[\r\n]|[^a])`. We also have to take into account whether any
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:502]: Do we need to worry about "a" being (possibly escaped) carriage-return or newline characters in the example for `[^a]`?

Comment (andygrove) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:502]: Nice catch... I added some more tests and they did expose a bug with the input pattern `[^\r\n]`.

**Comment by andygrove** [comment]
build

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Pull request merged**